[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86243

FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

fegin committed:
[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager

FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2059,6 +2053,46 @@ def _get_training_state(
+
+        prev_state_dict_type = None
+        prev_state_dict_config = None
+        # Use default config a state_dict config is not set.
```
"_if_ a state_dict config is not set"

rohan-varma reviewed (approved):
LGTM, just looks like there is a valid lint. Thanks for adding this!

fegin committed:
Update on "[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager"


FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager"


FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2111,6 +2104,46 @@ def _get_training_state(
         return next(iter(training_states))
 
+    @staticmethod
+    def set_state_dict_type(
```
Should we add a docstring?

fegin committed:
Update on "[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager"


FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager"


FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager"


FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager"


FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager"


FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin commented:
@pytorchbot merge -f "failing tests are not related"
